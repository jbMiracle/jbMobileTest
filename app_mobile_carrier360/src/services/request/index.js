import request from './request.service';

export default request;
export * from './request.service';
