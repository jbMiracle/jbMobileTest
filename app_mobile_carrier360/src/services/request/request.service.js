import omit from 'lodash/omit';
import get from 'lodash/get';
import Base64 from 'Base64';
import { Platform } from 'react-native';
import Config from 'react-native-config';

/* A Function that will handling building and firing off a http request.
 * @param {Object} requestOptions - Used to build the request.
 * @return {Promise}
 */
/*
 * requestOptions Fields:
 * @param {String} url - *Required* the url to be used for the request
 * @param {String} method - *Required* the http method to use
 * @param {Object} headers - any additional items that need to be added to the header
 * @param {Boolean} omitAuth - used to stop username and password from being added to the header. *If this is undefined, auth WILL BE added to the header*
 */
export default function request(requestOptions) {
  if (!requestOptions.method || !requestOptions.url) {
    throw new Error('RequestOptions missing required property method or url');
  }

  const options = omit(requestOptions, ['omitAuth', 'headers']);
  const defaultErrMsg = `${requestOptions.method} request error for ${requestOptions.url}`;

  return fetch({ ...options, headers: combineHeaders(requestOptions) })
    .catch(err => get(err, 'message', defaultErrMsg))
    .then(response => response.json()
      .then((body) => {
        if (!response.ok) {
          throw Error(`Service returned ${response.status}\n calling ${response.url}\nwith headers ${JSON.stringify(response.headers)}\nand message: ${JSON.stringify(body)}`);
        }
        return body;
      }));
}

function setDefaultHeaders(headers) {
  const { APP_NAME } = Config;

  return {
    'Accept-Encoding': 'gzip,deflate',
    'Accept-Language': 'en-US',
    'Content-Type': 'application/json',
    'User-Agent': `${APP_NAME} (compatible; Mozilla/5.0; MSIE 9.0; Trident/5.0; ${Platform.OS})`,
    'Proxy-Connection': 'Keep-Alive',
    Connection: 'Keep-Alive',
    Accept: 'application/hal+json, application/json',
    ...headers,
  };
}

function combineHeaders(requestOptions) {
  // TODO: There are other items that will eventually need to be added to the header, like impersonation items.
  const optionHeaders = requestOptions.headers || {};

  if (requestOptions.omitAuth === undefined || !requestOptions.omitAuth) {
    return setDefaultHeaders({ ...optionHeaders, ...getCredentials() });
  }

  return setDefaultHeaders(optionHeaders);
}

function getCredentials() {
  // TODO: add ability to get username and password from wherever we decide to store it.
  const username = 'a';
  const password = 'b';
  const base = `Basic ${Base64.btoa(`${username}:${password}`)}`;
  return { Authorization: base };
}

export { setDefaultHeaders, combineHeaders };
