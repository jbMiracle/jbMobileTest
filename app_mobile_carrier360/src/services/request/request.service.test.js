import request, { combineHeaders } from '.';

const requestOptions = {
  url: 'http://example.com',
  method: 'GET',
  body: 'baconbaconbacon',
  headers: {
    BreakfastFoods: 'AreTheBest',
  },
};

const requestOptionsNoAuth = {
  ...requestOptions,
  omitAuth: true,
};

jest.mock('Platform', () => {
  const Platform = require.requireActual('Platform');
  Platform.OS = 'ios';

  return Platform;
});
describe('Request Service Header Handling', () => {
  it('combineHeaders() can combine optional and default headers', () => {
    const result = combineHeaders(requestOptions);

    expect(result).toMatchSnapshot();
  });

  it('combineHeaders() can omit auth from headers when necessary', () => {
    const resultWithAuth = combineHeaders(requestOptions);
    const resultWithOutAuth = combineHeaders(requestOptionsNoAuth);

    expect(resultWithOutAuth).not.toEqual(resultWithAuth);
    expect(resultWithOutAuth).toMatchSnapshot();
  });
});

describe('Request Service Request Handling', () => {
  it('request() can make an HTTP request using fetch', async () => {
    const json = jest.fn().mockImplementation(options => Promise.resolve({ ...options }));
    const res = options => ({ ok: true, json: () => json(options) });
    global.fetch = jest.fn().mockImplementation(options => Promise.resolve(res(options)));

    const response = await request(requestOptions);
    expect(response).toMatchSnapshot();
  });

  it('request() can throw when missing required parameters', () => {
    expect(() => request({})).toThrow('RequestOptions missing required property method or url');
  });

  it('request() can return an error message when an http request fails', () => {
    global.fetch = jest.fn().mockImplementation(() => Promise.reject());

    return request({ url: 'http://example.com', method: 'POST' })
      .catch(err => expect(err).toMatchSnapshot());
  });
});
