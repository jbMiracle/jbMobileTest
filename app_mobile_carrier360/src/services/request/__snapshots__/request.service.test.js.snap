// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Service Header Handling combineHeaders() can combine optional and default headers 1`] = `
Object {
  "Accept": "application/hal+json, application/json",
  "Accept-Encoding": "gzip,deflate",
  "Accept-Language": "en-US",
  "Authorization": "Basic YTpi",
  "BreakfastFoods": "AreTheBest",
  "Connection": "Keep-Alive",
  "Content-Type": "application/json",
  "Proxy-Connection": "Keep-Alive",
  "User-Agent": "carrier360 (compatible; Mozilla/5.0; MSIE 9.0; Trident/5.0; ios)",
}
`;

exports[`Request Service Header Handling combineHeaders() can omit auth from headers when necessary 1`] = `
Object {
  "Accept": "application/hal+json, application/json",
  "Accept-Encoding": "gzip,deflate",
  "Accept-Language": "en-US",
  "BreakfastFoods": "AreTheBest",
  "Connection": "Keep-Alive",
  "Content-Type": "application/json",
  "Proxy-Connection": "Keep-Alive",
  "User-Agent": "carrier360 (compatible; Mozilla/5.0; MSIE 9.0; Trident/5.0; ios)",
}
`;

exports[`Request Service Request Handling request() can make an HTTP request using fetch 1`] = `
Object {
  "body": "baconbaconbacon",
  "headers": Object {
    "Accept": "application/hal+json, application/json",
    "Accept-Encoding": "gzip,deflate",
    "Accept-Language": "en-US",
    "Authorization": "Basic YTpi",
    "BreakfastFoods": "AreTheBest",
    "Connection": "Keep-Alive",
    "Content-Type": "application/json",
    "Proxy-Connection": "Keep-Alive",
    "User-Agent": "carrier360 (compatible; Mozilla/5.0; MSIE 9.0; Trident/5.0; ios)",
  },
  "method": "GET",
  "url": "http://example.com",
}
`;

exports[`Request Service Request Handling request() can return an error message when an http request fails 1`] = `[TypeError: response.json is not a function]`;
